presto-main/src/main/java/io/prestosql/operator/exchange/LocalExchangeMemoryManager.java

```diff
@@ -47,31 +45,39 @@ public LocalExchangeMemoryManager(long maxBufferedBytes)
 
     public void updateMemoryUsage(long bytesAdded)
     {
-        SettableFuture<?> future;
-        synchronized (this) {
-            bufferedBytes.addAndGet(bytesAdded);
-
-            // if we are full, then breakout
-            if (bufferedBytes.get() > maxBufferedBytes || notFullFuture.isDone()) {
-                return;
+        long bufferedBytes = this.bufferedBytes.addAndGet(bytesAdded);
+        // detect the transition from above to below the full boundary
+        if (bufferedBytes <= maxBufferedBytes && (bufferedBytes - bytesAdded) > maxBufferedBytes) {
+            SettableFuture<?> future;
+            synchronized (this) {
+                // if we have no callback waiting, return early
```

Comment:
This optimization reduces readability of concurrent code for little benefit. In practice, we already synchronized on `this`, so doing extra checks like `notFullFuture.isDone()` is not a big issue. Also, we should recheck that buffer is still not full before triggering future. That check is cheap, but waking up drivers to do empty processing is not.

> The effect was that previously the transition edge from above-full to below full always unblocked any waiters, and not doing so here (even if we've transitioned back up above the full boundary in the mean time) is a departure from that behavior.

We cold not transition back to above in the meantime because all methods in `LocalExchangeMemoryManager` are synchronized. We could transition to above immediately after future was set, but that's slightly different behavior.

Reply:
> This optimization reduces readability of concurrent code...

I disagree. The prior implementation seemed "simpler" in that there was no null checking, but it was also masking a subtle problem: checking `isDone()` while holding the lock can race with `future.set(null)` which occurs _outside_ of the synchronized section. This happens to be _ok_ because the field is overwritten inside the synchronized section, but is much more subtle than a null check IMO.

> ... so doing extra checks like `notFullFuture.isDone()` is not a big issue

It's significantly more expensive than a null check because it requires a volatile read (while holding the lock) of the value at the other end of the current reference, meaning that having acquired the object monitor is insufficient to have brought that memory into the CPU cache and will _at best_ need to read it from a shared CPU cache if not from main memory.

Now, will the extra cycles incurred in doing that dominate performance? Not likely, but there is an Amdahl's law component to the duration that any thread spends in the synchronized section of `updateMemoryUsage(long)` or `getNotFullFuture()`.

> That check is cheap, but waking up drivers to do empty processing is not.

I agree, the question is whether changing the semantics are a good idea as part of this refactor. Previously, all modifications were linearized by the object monitor. Now, they linearize on the `bufferedBytes` atomic to accomplish the same goal and the synchronized block is establishing a consistent view of the callback field. I can't say how many workloads today depend on "fluttering" against the full vs not-full state and whether they would all benefit from having some of those unblocking notifications suppressed. In the interest of being conservative, I would prefer to keep the previous semantics (but with less lock contention) than potentially regress performance.